scrapy/utils/url.py

```diff
@@ -114,10 +115,10 @@ def escape_ajax(url):
 
 def add_http_if_no_scheme(url):
     """Add http as the default scheme if it is missing from the url."""
-    if url.startswith('//'):
-        url = 'http:' + url
-        return url
-    parser = parse_url(url)
-    if not parser.scheme or not parser.netloc:
-        url = 'http://' + url
+    match = re.match(r"^\w+://", url, flags=re.I)
+    parts = urlparse(url)
+    if not match:
+        scheme = "http:" if parts.netloc else "http://"
+        url = scheme + url
```

Comment:
@kmike , I haven't checked that part of the change.
`add_http_if_no_scheme` seems to be used only for [scrapy shell](https://github.com/scrapy/scrapy/blob/a97f28fbf09d2e6d97b626433630df8ca1931b3b/scrapy/commands/shell.py#L50)
[Tests](https://github.com/scrapy/scrapy/blob/9676cf8e53affbc9f2dc95a8c3e14601e8e2bfdb/tests/test_utils_url.py#L193) look thorough enough for usual cases

Still, I will move `parts = urlparse(url)` inside `if not match:` branch at least